interp/src/environment.rs

```diff
@@ -272,34 +289,69 @@ impl InterpreterState {
     /// from the source's stack environment allowing divergence from the fork
     /// point
     pub fn fork(&mut self) -> Self {
-        let other_pv_map = if self.pv_map.top().is_empty() {
-            self.pv_map.fork_from_tail()
+        let other_pv_map = if self.port_map.top().is_empty() {
+            self.port_map.fork_from_tail()
         } else {
-            self.pv_map.fork()
+            self.port_map.fork()
         };
         Self {
             clk: self.clk,
-            cell_prim_map: Rc::clone(&self.cell_prim_map),
-            pv_map: other_pv_map,
+            cell_map: self.cell_map.clone(),
+            port_map: other_pv_map,
             context: Rc::clone(&self.context),
+            comp_name: self.comp_name.clone(),
+        }
+    }
+    /// Creates a fork of the source environment which has the same clock and
+    /// underlying primitive map but whose stack environment has been forked
+    /// from the source's stack environment allowing divergence from the fork
+    /// point. This forces the creation of a new layer, unlike fork
+    pub fn force_fork(&mut self) -> Self {
+        Self {
+            clk: self.clk,
+            cell_map: self.cell_map.clone(),
+            port_map: self.port_map.fork(),
+            context: Rc::clone(&self.context),
+            comp_name: self.comp_name.clone(),
         }
     }
 
     /// Merge the given environments. Must be called from the root environment
-    pub fn merge_many(mut self, others: Vec<Self>) -> Self {
+    pub fn merge_many(
+        mut self,
+        others: Vec<Self>,
+        overlap: &HashSet<*const ir::Port>,
+    ) -> InterpreterResult<Self> {
         let clk = others
             .iter()
             .chain(once(&self))
             .map(|x| x.clk)
             .max()
             .unwrap(); // safe because of once
 
-        self.pv_map = self
-            .pv_map
-            .merge_many(others.into_iter().map(|x| x.pv_map).collect());
+        let port_map = self.port_map;
+        let merged = port_map.merge_many(
+            others.into_iter().map(|x| x.port_map).collect(),
+            overlap,
+        );
+
+        self.port_map = match merged {
+            Ok(ok) => Ok(ok),
+            Err(e) => {
+                let mut ie: InterpreterError = e.into();
+                if let InterpreterError::ParOverlap { parent_id, .. } = &mut ie
```

Comment:
Is this the right place to catch and modify this error? For example, is it not possible for a set of continuous assignments to also assign to the component's `done` signal? Is that handled by a different error message?

Reply:
It is possible for that to happen. This error doesn't occur at the point of overlapping assignments, it only happens during the merger of the various par arms.

You're correct about the continuous assignments, I'm not entirely sure why that hasn't become an issue yet. Most likely anything assigned by a continuous assignment should be added to the overlap list.